### DISCUSSION TITLE
Error message in query execution log

### DISCUSSION THREAD

**Discussion by balhoff** [comunica/comunica#1279] [Q&A]
Running a SPARQL query against a TPF server using the Comunica web UI, I see this within the execution log:

```
Expression error for extend operation with bindings '{
  "gender": "http://www.wikidata.org/entity/Q43445",
  "birthdate": "\"1954-01-01T00:00:00Z\"^^http://www.w3.org/2001/XMLSchema#dateTime",
  "father": "http://www.wikidata.org/entity/Q100355207",
  "horse": "http://www.wikidata.org/entity/Q100355482",
  "var0": "http://www.wikidata.org/entity/Q726",
  "birthyear": "\"1954\"^^http://www.w3.org/2001/XMLSchema#integer"
}' {"actor":"urn:comunica:default:query-operation/actors#extend"}
```

Does this indicate a problem with the server? Or is is just part of the normal query execution?

**Comment by rubensworks** [comment]
This kind of error indicates that the extend operation is failing on a specific binding, after which it is skipped from the final results. It's part of normal execution as prescribed by the SPARQL spec.

I'm surprised the actual error message isn't shown though...

@jitsedesmet Perhaps you have some additional insights?

**Reply by jitsedesmet** [reply]
I also don't know without the error message. I checked and both the `dateTime` and `integer` have a correct format. So no low hanging fruits there.

**Reply by rubensworks** [reply]
So ideally, we'd log the actual error message here as well: https://github.com/comunica/comunica/blob/master/packages/actor-query-operation-extend/lib/ActorQueryOperationExtend.ts#L57

**Reply by jitsedesmet** [reply]
Your wish is my command https://github.com/comunica/comunica/pull/1280

**Reply by balhoff** [reply]
Thanks for the info @rubensworks. I compared with a SPARQL endpoint for the same dataset, and the total number of results seems to be the same, so I think it is working correctly. The query looks like this:

```sparql
PREFIX wd: <http://www.wikidata.org/entity/>
PREFIX wdt: <http://www.wikidata.org/prop/direct/>
PREFIX wikibase: <http://wikiba.se/ontology#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX schema: <http://schema.org/>
SELECT DISTINCT ?horse ?mother ?father (year(?birthdate) as ?birthyear) (year(?deathdate) as ?deathyear)
WHERE
{
  ?horse wdt:P31/wdt:P279* wd:Q726 .     # Instance et sous-classes de Q726-Cheval
   
  OPTIONAL{?horse wdt:P25 ?mother .}       # P25  : Mère
  OPTIONAL{?horse wdt:P22 ?father .}       # P22  : Père
  OPTIONAL{?horse wdt:P569 ?birthdate .} # P569 : Date de naissance
  OPTIONAL{?horse wdt:P570 ?deathdate .}     # P570 : Date de décès
  OPTIONAL{?horse wdt:P21 ?gender .}       # P21  : Sexe
}
ORDER BY ?horse
```

The output seems to be caused by the `YEAR` functions in the `SELECT` line (I removed those and didn't get the messages). And I guess for the horse in the error message (`Q100355482`) `?deathdate` is unbound so maybe that's the source of the output. But since as you say that's just normal evaluation of the SPARQL query, it's confusing that this message is printed.

**Reply by rubensworks** [reply]
The SPARQL spec does not prescribe such a message to be emitted.
We decided to emit it because some users were getting confused of some results with expressions somehow silently being removed (due to expression errors).
So in order to improve debugging of such queries (if that would be desired), we added this message.

If you have a suggestion of how we could reduce confusion on this matter, we're certainly open to it :-)

**Reply by balhoff** [reply]
Thanks, now that I know what's happening I won't worry that I'm missing results. Perhaps the wording could be revised to clarify that  the "error" is in the SPARQL sense of applying a function to an unbound value, not a software error. But I'm not sure how many other new users would have the same confusion.
